torch/distributions/truncated_distribution.py

```diff
@@ -0,0 +1,67 @@
+import torch
+from torch.distributions import constraints
+from torch.distributions.distribution import Distribution
+from torch.distributions.utils import broadcast_all
+
+
+class TruncatedDistribution(Distribution):
+    r"""
+    Extension of the Distribution class, which applies truncation to a base distribution.
+    """
+
+    has_rsample = True
+
+    def __init__(self, base_distribution, lower_bound=-float('inf'), upper_bound=float('inf'), *args, **kwargs):
+        super(TruncatedDistribution, self).__init__(*args, **kwargs)
```

Comment:
I'm pretty sure we'll need to compute `batch_shape` and `event_shape` by broadcasting `base_distribution.batch_shape+base_distribution.event_shape` with `lower_bound` and `upper_bound` shapes.